db.go

```diff
@@ -501,25 +528,35 @@ func (db *DB) syncWALToLTX(ctx context.Context, ltxFilename string) error {
 		return err
 	}
 
-	// If salt matches the salt in the LTX file & the WAL size exceeds the
-	// offset in the LTX file then we need to truncate the WAL to match.
-	// This occurs before the checkpoint() call so we don't replay too much.
+	// If WAL salt doesn't match the LTX WAL salt then the WAL has been
+	// restarted and we need to remove it. We are just renaming it for now so
+	// we can debug in case this happens.
 	salt1 := binary.BigEndian.Uint32(hdr[16:])
 	salt2 := binary.BigEndian.Uint32(hdr[20:])
 	if salt1 != r.Header().WALSalt1 || salt2 != r.Header().WALSalt2 {
-		log.Printf("wal salt mismatch on %q, skipping sync with ltx", db.name)
-		return nil
-	} else if fi.Size() <= ltxWALSize {
-		log.Printf("wal size in sync on %q, skipping truncation", db.name)
+		log.Printf("wal salt mismatch on %q, removing wal", db.name)
+		if err := os.Rename(db.WALPath(), db.WALPath()+".removed"); err != nil {
+			return fmt.Errorf("rename wal file with salt mismatch: %w", err)
+		}
 		return nil
 	}
```

Comment:
This fixes a bug where salt mismatch caused the WAL file to remain. This means it would be replayed by the SQLite client on recovery and corrupt the database. Now, if the salt doesn't match then it renames the file with a `.removed` suffix so it can be inspected later, if need be.